<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="statgenQTLxT">
<title>Multi-trait GWAS with the statgenQTLxT package • statgenQTLxT</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multi-trait GWAS with the statgenQTLxT package">
<meta property="og:description" content="statgenQTLxT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">statgenQTLxT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/index.html">Vignettes</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-file-code-o"></span>
     
    functions
  </a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--other-packages">
    <span class="fa fa-seedling"></span>
     
    other packages
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--other-packages">
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenSTA">statgenSTA</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenGxE">statgenGxE</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenGWAS">statgenGWAS</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenHTP">statgenHTP</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenIBD">statgenIBD</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/statgenMPP">statgenIBD</a>
    <a class="external-link dropdown-item" href="https://biometris.github.io/LMMsolver">LMMsolver</a>
  </div>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Biometris/statgenQTLxT">
    <span class="fa fa-github fa-lg fab"></span>
     
    github
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Multi-trait GWAS with the statgenQTLxT package</h1>
                        <h4 data-toc-skip class="author">Bart-Jan van
Rossum</h4>
            
            <h4 data-toc-skip class="date">2023-02-15</h4>
      
      
      <div class="d-none name"><code>statgenQTLxT.Rmd</code></div>
    </div>

    
    
<p>The <code>statgenQTLxT</code> package performs multi-trait and
multi-environment Genome Wide Association Studies (GWAS), following the
approach of <span class="citation">(<a href="#ref-Zhou2014" role="doc-biblioref">Zhou and Stephens 2014</a>)</span>. It builds on
the <code>statgenGWAS</code> package (for single trait GWAS) which is
available from <a href="https://biometris.github.io/statgenGWAS/" class="external-link">CRAN</a>. The package
uses data structures and plots defined in the <code>statgenGWAS</code>
package. It is recommended to read the vignette of this package,
accessible in R via <code>vignette(package = "statgenGWAS")</code> or
online at <a href="https://biometris.github.io/statgenGWAS/articles/GWAS.html" class="external-link uri">https://biometris.github.io/statgenGWAS/articles/GWAS.html</a>
to get a general idea of those.</p>
<hr>
<div class="section level2">
<h2 id="multi-trait-gwas">Multi-Trait GWAS<a class="anchor" aria-label="anchor" href="#multi-trait-gwas"></a>
</h2>
<div class="section level3">
<h3 id="theoretical-background">Theoretical background<a class="anchor" aria-label="anchor" href="#theoretical-background"></a>
</h3>
<p>Multi-trait GWAS in the <code>statgenQTLxT</code> package estimates
and tests the effect of a SNP in different trials or on different
traits, one SNP at a time. Genetic and residual covariances are fitted
only once, for a model without SNPs. Given balanced data on <span class="math inline">\(n\)</span> genotypes and <span class="math inline">\(p\)</span> traits (or trials) we fit a mixed model
of the form</p>
<p><span class="math inline">\(Y = \left(\begin{array}{c} Y_1 \\ \vdots
\\ Y_p\end{array}\right) = \left(\begin{array}{c} X\gamma_1 \\ \vdots \\
X\gamma_p\end{array}\right) + \left(\begin{array}{c} x\beta_1 \\ \vdots
\\ x\beta_p\end{array}\right) + \left(\begin{array}{c} G_1 \\ \vdots \\
G_p\end{array}\right) + \left(\begin{array}{c} E_1 \\ \vdots \\
E_p\end{array}\right)\)</span></p>
<p>where <span class="math inline">\(Y_1\)</span> to <span class="math inline">\(Y_p\)</span> are <span class="math inline">\(n ×
1\)</span> vectors with the phenotypic values for traits or trials <span class="math inline">\(1, \ldots, p\)</span>. <span class="math inline">\(x\)</span> is the <span class="math inline">\(n ×
1\)</span> vector of scores for the marker under consideration, and
<span class="math inline">\(X\)</span> the <span class="math inline">\(n
× q\)</span> design matrix for the other covariates. By default only a
trait (environment) specific intercept is included. The vector
containing the genetic background effects <span class="math inline">\(G_1, \ldots, G_p\)</span> is Gaussian with zero
mean and covariance <span class="math inline">\(V_g \otimes K\)</span>,
where <span class="math inline">\(V_g\)</span> is a <span class="math inline">\(p × p\)</span> matrix of genetic (co)variances,
and <span class="math inline">\(K\)</span> an <span class="math inline">\(n × n\)</span> kinship matrix. Similarly, the
residual errors (<span class="math inline">\(E_1, \ldots, E_p\)</span>)
have covariance <span class="math inline">\(V_e \otimes I_n\)</span>,
for a <span class="math inline">\(p × p\)</span> matrix <span class="math inline">\(V_e\)</span> of residual (co)variances.</p>
<div class="section level4">
<h4 id="common">Hypotheses for the SNP-effects<a class="anchor" aria-label="anchor" href="#common"></a>
</h4>
<p>For each SNP, the null-hypothesis <span class="math inline">\(\beta_1
= \dots = \beta_p = 0\)</span> is tested, using the likelihood ratio
test (LRT) described in <span class="citation">(<a href="#ref-Zhou2014" role="doc-biblioref">Zhou and Stephens 2014</a>)</span>. If
<code>estCom = TRUE</code>, additional tests for a common effect and for
QTL-by-trait (QTL × T) or QTL-by-environment (QTL × E) are performed,
using the parameterization <span class="math inline">\(\beta_j = \alpha
+ \alpha_j (1 \leq j \leq p)\)</span>. As in <span class="citation">(<a href="#ref-Korte2012" role="doc-biblioref">Korte et al.
2012</a>)</span>, we use likelihood ratio tests, but not restricted to
the bivariate case. For the common effect, we fit the reduced model
<span class="math inline">\(\beta_j = \alpha\)</span>, and test if <span class="math inline">\(\alpha = 0\)</span>. For the interactions, we test
if <span class="math inline">\(\alpha_1 = \dots = \alpha_p =
0\)</span>.</p>
</div>
<div class="section level4">
<h4 id="vc">Models for the genetic and residual covariance<a class="anchor" aria-label="anchor" href="#vc"></a>
</h4>
<p><span class="math inline">\(V_g\)</span> and <span class="math inline">\(V_e\)</span> can be provided by the user
(<code>fitVarComp = FALSE</code>); otherwise one of the following models
is used, depending on <code>covModel</code>. If
<code>covModel = "unst"</code>, an unstructured model is assumed, as in
<span class="citation">(<a href="#ref-Zhou2014" role="doc-biblioref">Zhou and Stephens 2014</a>)</span>: <span class="math inline">\(V_g\)</span> and <span class="math inline">\(V_e\)</span> can be any positive-definite matrix,
requiring a total of <span class="math inline">\(p(p+1)/2\)</span>
parameters per matrix. If <code>covModel = "fa"</code>, a
factor-analytic model is fitted using an EM-algorithm, as in <span class="citation">(<a href="#ref-Millet2016" role="doc-biblioref">Millet
et al. 2016</a>)</span>. <span class="math inline">\(V_g\)</span> and
<span class="math inline">\(V_e\)</span> are assumed to be of the form
<span class="math inline">\(W W^t + D\)</span>, where <span class="math inline">\(W\)</span> is a <span class="math inline">\(p ×
m\)</span> matrix of factor loadings and <span class="math inline">\(D\)</span> a diagonal matrix with trait or
environment specific values. <span class="math inline">\(m\)</span> is
the order of the model, and the arguments <code>mG</code> and
<code>mE</code> specify the order used for respectively <span class="math inline">\(V_g\)</span> and <span class="math inline">\(V_e\)</span>. <code>maxIter</code> sets the
maximum number of iterations used in the EM-algorithm. Finally, if
<code>covModel = "pw"</code>, <span class="math inline">\(V_g\)</span>
and <span class="math inline">\(V_e\)</span> are estimated ‘pairwise’,
as in <span class="citation">(<a href="#ref-Furlotte2015" role="doc-biblioref">Furlotte and Eskin 2015</a>)</span>. Looping over
pairs of traits or trials <span class="math inline">\(1 \leq j &lt; k
\leq p\)</span>, <span class="math inline">\(V_g[j,k] =
V_g[k,j]\)</span> and <span class="math inline">\(V_e[j,k] =
V_e[k,j]\)</span> are estimated assuming a bivariate mixed model. The
diagonals of <span class="math inline">\(V_g\)</span> and <span class="math inline">\(V_e\)</span> are fitted assuming univariate mixed
models. If the resulting <span class="math inline">\(V_g\)</span> or
<span class="math inline">\(V_e\)</span> is not positive-definite, they
are replaced by the nearest positive-definite matrix. In case
<code>covModel = "unst"</code> or <code>"pw"</code> it is possible to
assume that <span class="math inline">\(V_e\)</span> is diagonal
(<code>VeDiag = TRUE</code>).</p>
<hr>
</div>
</div>
</div>
<div class="section level2">
<h2 id="the-class-gdata">The class <code>gData</code><a class="anchor" aria-label="anchor" href="#the-class-gdata"></a>
</h2>
<p>Data for analysis on genomic data comes from different sources and is
stored in one data object of class <strong>gData</strong>
(<strong>g</strong>enomic <strong>Data</strong>) for convenience. A
<code>gData</code> object will contain all data needed for performing
analyses, so the first thing to do when using the
<code>statgenQTLxT</code> package is creating a <code>gData</code>
object; see the <code>statgenGWAS</code> package for more details. A
<code>gData</code> object contains the following components:</p>
<ul>
<li>Marker map, a data.frame describing the physical positions of the
markers on the chromosomes,</li>
<li>Marker matrix, a numerical matrix containing the genotyping,</li>
<li>Phenotypic data, either a single data.frame or a list of data.frames
containing the phenotypic data,</li>
<li>Kinship matrix, a matrix describing the genetic relatedness between
the different genotypes,</li>
<li>Further covariates that can be used in the analyses.</li>
</ul>
<p>In our examples below we will show how a <code>gData</code> object is
created.</p>
<hr>
</div>
<div class="section level2">
<h2 id="worked-example-1-multiple-traits-in-one-trial">Worked example 1: multiple traits in one trial<a class="anchor" aria-label="anchor" href="#worked-example-1-multiple-traits-in-one-trial"></a>
</h2>
<p>As examples of the functionality of the package two worked examples
are provided using maize data from the European Union project DROPS. The
data is available from <a href="https://doi.org/10.15454/IASSTN" class="external-link uri">https://doi.org/10.15454/IASSTN</a> <span class="citation">(<a href="#ref-Millet2019" role="doc-biblioref">Millet
et al. 2019</a>)</span> and the relevant data sets are included as
data.frames in the <code>statgenGWAS</code> package. We will first show
how to load the data and create a <code>gData</code> object. Users
already familiar with the <code>statgenGwas</code> packages might want
to skip this part and go straight to <a href="#mtg"><strong>Running
Multi-trait GWAS</strong></a> section.</p>
<p><code>dropsMarkers</code> contains the coded marker information for
41,722 SNPs and 246 genotypes. <code>dropsMap</code> contains
information about the positions of those SNPs on the B73 reference
genome V2. <code>dropsPheno</code> contains data for the genotypic means
(Best Linear Unbiased Estimators, BLUEs) for a subset of ten
experiments, with one value per experiment per genotype, for eight
traits. For a more detailed description of the contents of the data see
<code><a href="https://rdrr.io/pkg/statgenGWAS/man/dropsData.html" class="external-link">help(dropsData, package = statgenGWAS)</a></code>.</p>
<div class="section level3">
<h3 id="create-gdata-object">
<strong>Create gData object</strong><a class="anchor" aria-label="anchor" href="#create-gdata-object"></a>
</h3>
<p>The first step is to create a <strong><code>gData</code></strong>
object from the raw data that can be used for the GWAS analysis. For
this the raw data has to be converted to a suitable format for a
<strong><code>gData</code></strong> object, see
<code><a href="https://rdrr.io/pkg/statgenGWAS/man/gData.html" class="external-link">help(createGData, package = statgenGWAS)</a></code> and the
<code>statgenGWAS</code> vignette for more details.</p>
<p>When running a multi-trait or multi-environment GWAS, all traits used
in the analysis should be in the same <code>data.frame</code>, with
genotype as first column and the phenotypic data in subsequent columns.
In case of a multi-trait analysis the phenotypic columns contain
different traits, measured in one environment, while for a
multi-environment the columns correspond to the different environments
(same trait). In both cases the <code>data.frame</code> may only contain
phenotypic data. Additional covariates need to be stored in
<code>covar</code>.</p>
<p>Below are some examples of what these <code>data.frames</code> should
look like.</p>
<table class="table">
<thead><tr class="header">
<th align="left">genotype</th>
<th align="left">Trait1</th>
<th align="left">Trait2</th>
<th align="left">Trait3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">G1</td>
<td align="left">0.3</td>
<td align="left">17</td>
<td align="left">277</td>
</tr>
<tr class="even">
<td align="left">G2</td>
<td align="left">0.4</td>
<td align="left">19</td>
<td align="left">408</td>
</tr>
<tr class="odd">
<td align="left">G3</td>
<td align="left">0.5</td>
<td align="left">17</td>
<td align="left">206</td>
</tr>
<tr class="even">
<td align="left">G4</td>
<td align="left">0.7</td>
<td align="left">13</td>
<td align="left">359</td>
</tr>
</tbody>
</table>
<table class="table">
<thead><tr class="header">
<th align="left">genotype</th>
<th align="left">Trait1-Trial1</th>
<th align="left">Trait1-Trial2</th>
<th align="left">Trait1-Trial3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">G1</td>
<td align="left">0.3</td>
<td align="left">0.7</td>
<td align="left">0.5</td>
</tr>
<tr class="even">
<td align="left">G2</td>
<td align="left">0.4</td>
<td align="left">0.9</td>
<td align="left">0.1</td>
</tr>
<tr class="odd">
<td align="left">G3</td>
<td align="left">0.5</td>
<td align="left">0.8</td>
<td align="left">0.2</td>
</tr>
<tr class="even">
<td align="left">G4</td>
<td align="left">0.7</td>
<td align="left">0.5</td>
<td align="left">0.4</td>
</tr>
</tbody>
</table>
<p>In our first example, we want to perform multi-trait GWAS for one of
the DROPS environments. <code>dropsPheno</code> contains genotypic means
for six traits in ten trials. To run a multi-trait GWAS analysis for
each of the ten trials, the data has to be added as a <code>list</code>
of ten <code>data.frames</code>. Recall that these
<code>data.frames</code> should have “genotype” as their first column
and may only contain traits after that. Other columns need to be
dropped.</p>
<p>The code below creates this <code>list</code> of
<code>data.frames</code> from <code>dropsPheno</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">dropsPheno</span>, package <span class="op">=</span> <span class="st">"statgenGWAS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Convert phenotypic data to a list.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"genotype"</span></span>
<span><span class="va">dropsPheno</span> <span class="op">&lt;-</span> <span class="va">dropsPheno</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Experiment"</span>, <span class="st">"genotype"</span>, <span class="st">"grain.yield"</span>, <span class="st">"grain.number"</span>,</span>
<span>                           <span class="st">"anthesis"</span>, <span class="st">"silking"</span>, <span class="st">"plant.height"</span>, <span class="st">"ear.height"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## Split data by experiment.</span></span>
<span><span class="va">dropsPhenoList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dropsPheno</span>, f <span class="op">=</span> <span class="va">dropsPheno</span><span class="op">[[</span><span class="st">"Experiment"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## Remove Experiment column.</span></span>
<span><span class="co">## phenotypic data should consist only of genotype and traits.</span></span>
<span><span class="va">dropsPhenoList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">dropsPhenoList</span>, FUN <span class="op">=</span> <span class="va">`[`</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>If the phenotypic data consists of only one trial/experiment, it can
be added as a single <code>data.frame</code> without first converting it
to a <code>list</code>. In that case <code>createGData</code> will
convert the input to a <code>list</code> with one item.</p>
<p>Now a <code>gData</code> object containing map, marker information,
and phenotypes can be created. Kinship matrix and covariates may be
added later on.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load marker data.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dropsMarkers"</span>, package <span class="op">=</span> <span class="st">"statgenGWAS"</span><span class="op">)</span></span>
<span><span class="co">## Add genotypes as row names of dropsMarkers and drop Ind column.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMarkers</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[[</span><span class="st">"Ind"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">dropsMarkers</span> <span class="op">&lt;-</span> <span class="va">dropsMarkers</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Load genetic map.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"dropsMap"</span>, package <span class="op">=</span> <span class="st">"statgenGWAS"</span><span class="op">)</span></span>
<span><span class="co">## Add genotypes as row names of dropsMap.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dropsMap</span><span class="op">[[</span><span class="st">"SNP.names"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## Rename Chromosome and Position columns.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dropsMap</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">"pos"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create a gData object containing map, marker and phenotypic information.</span></span>
<span><span class="va">gDataDrops</span> <span class="op">&lt;-</span> <span class="fu">statgenGWAS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/statgenGWAS/man/gData.html" class="external-link">createGData</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">dropsMarkers</span>,</span>
<span>                                       map <span class="op">=</span> <span class="va">dropsMap</span>, </span>
<span>                                       pheno <span class="op">=</span> <span class="va">dropsPhenoList</span><span class="op">)</span></span></code></pre></div>
<p>To get an idea of the contents of the data a summary of the
<code>gData</code> object can be made. This will give an overview of the
content of the map and markers and also print a summary per trait per
trial. Since there are ten trials and six traits in
<code>gDataDrops</code> we restrict the output to one trial, using the
<code>trials</code> argument of the summary function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Summarize gDataDrops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gDataDrops</span>, trials <span class="op">=</span> <span class="st">"Mur13W"</span><span class="op">)</span></span>
<span><span class="co">#&gt; map</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of chromosomes: 10 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; markers</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of genotypes: 246 </span></span>
<span><span class="co">#&gt;  Content:</span></span>
<span><span class="co">#&gt;   0 1 2 NA </span></span>
<span><span class="co">#&gt;   0.28 0.01 0.71 0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; pheno</span></span>
<span><span class="co">#&gt;  Number of trials: 1 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Mur13W:</span></span>
<span><span class="co">#&gt;      Number of traits: 6 </span></span>
<span><span class="co">#&gt;      Number of genotypes: 246 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;         grain.yield grain.number anthesis silking plant.height ear.height</span></span>
<span><span class="co">#&gt; Min.            3.3         1348       56      59          222        102</span></span>
<span><span class="co">#&gt; 1st Qu.         6.3         2641       61      64          251        125</span></span>
<span><span class="co">#&gt; Median          7.5         2965       63      66          258        133</span></span>
<span><span class="co">#&gt; Mean            7.4         2986       63      66          259        133</span></span>
<span><span class="co">#&gt; 3rd Qu.         8.4         3359       66      68          266        141</span></span>
<span><span class="co">#&gt; Max.           11.4         4510       71      74          294        172</span></span>
<span><span class="co">#&gt; NA's            0.0            0        0       0            0          0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recoding-and-cleaning-of-markers">
<strong>Recoding and cleaning of markers</strong><a class="anchor" aria-label="anchor" href="#recoding-and-cleaning-of-markers"></a>
</h3>
<p>Marker data has to be numerical and without missing values in order
to do GWAS analysis. This can be achieved using the
<code>codeMarkers()</code> function, which can also perform imputation
of missing markers. The marker data available for the DROPS project has
already been converted from A/T/C/G to 0/1/2. We still use the
<code>codeMarkers()</code> function to further clean the markers, in
this case by removing the duplicate SNPs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set seed.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Remove duplicate SNPs from gDataDrops.</span></span>
<span><span class="va">gDataDropsDedup</span> <span class="op">&lt;-</span> <span class="fu">statgenGWAS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/statgenGWAS/man/codeMarkers.html" class="external-link">codeMarkers</a></span><span class="op">(</span><span class="va">gDataDrops</span>, impute <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                            verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span></span>
<span><span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 0 SNPs removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 5098 duplicate SNPs removed.</span></span>
<span><span class="co">#&gt; Output contains 36624 SNPs for 246 genotypes.</span></span></code></pre></div>
<p>Note that duplicate SNPs are removed at random. To get reproducible
results make sure to set a seed.</p>
<p>To demonstrate the options of the <code>codeMarkers()</code>
function, see <code><a href="https://rdrr.io/pkg/statgenGWAS/man/codeMarkers.html" class="external-link">help(codeMarkers, package = statgenGWAS)</a></code> and
the <code>statgenGWAS</code> vignette for more details.</p>
</div>
<div class="section level3">
<h3 id="mtg">
<strong>Running Multi-trait GWAS</strong><a class="anchor" aria-label="anchor" href="#mtg"></a>
</h3>
<p>The cleaned <code>gData</code> object can be used for performing
multi-trait GWAS analysis. In this example the trial <em>Mur13W</em> is
used to demonstrate the options of the <code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code>
function, for a subset of five traits. As in [millet2016] we choose a
factor analytic model for the genetic and residual covariance.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for 5 traits in trial Mur13W.</span></span>
<span><span class="va">GWASDrops</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>, </span>
<span>                               traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grain.yield"</span>,<span class="st">"grain.number"</span>,</span>
<span>                                          <span class="st">"anthesis"</span>, <span class="st">"silking"</span> ,<span class="st">"plant.height"</span><span class="op">)</span>,</span>
<span>                               trials <span class="op">=</span> <span class="st">"Mur13W"</span>, </span>
<span>                               covModel <span class="op">=</span> <span class="st">"fa"</span><span class="op">)</span></span></code></pre></div>
<p>The output of the <code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function is an
object of class <code>GWAS</code>. This is a list consisting of five
elements described below.<br></p>
<ol style="list-style-type: decimal">
<li>
<strong>GWAResult</strong>: a list of <code>data.tables</code>, one
for each trial for which the analysis was run. Each
<code>data.table</code> has the following columns:</li>
</ol>
<table class="table">
<colgroup>
<col width="14%">
<col width="85%">
</colgroup>
<tbody>
<tr class="odd">
<td align="left">snp</td>
<td align="left">SNP name</td>
</tr>
<tr class="even">
<td align="left">trait</td>
<td align="left">trait name</td>
</tr>
<tr class="odd">
<td align="left">chr</td>
<td align="left">chromosome on which the SNP is located</td>
</tr>
<tr class="even">
<td align="left">pos</td>
<td align="left">position of the SNP on the chromosome</td>
</tr>
<tr class="odd">
<td align="left">pValue</td>
<td align="left">P-value for the SNP</td>
</tr>
<tr class="even">
<td align="left">LOD</td>
<td align="left">LOD score for the SNP, defined as <span class="math inline">\(-\log_{10}(pValue)\)</span>
</td>
</tr>
<tr class="odd">
<td align="left">effect</td>
<td align="left">effect of the SNP on the trait value</td>
</tr>
<tr class="even">
<td align="left">effectSe</td>
<td align="left">standard error of the effect of the SNP on the trait
value</td>
</tr>
<tr class="odd">
<td align="left">allFreq</td>
<td align="left">allele frequency of the SNP</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">GWAResult</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">)</span></span>
<span><span class="co">#&gt;              snp    trait chr    pos pValue   LOD effect effectSe allFreq</span></span>
<span><span class="co">#&gt; 1:         SYN83 anthesis   1   3498  0.928 0.033  -0.14     0.19    0.60</span></span>
<span><span class="co">#&gt; 2: PZE-101000060 anthesis   1 157104  0.185 0.733   0.50     0.21    0.72</span></span>
<span><span class="co">#&gt; 3: PZE-101000088 anthesis   1 238347  0.078 1.110  -0.65     0.25    0.84</span></span>
<span><span class="co">#&gt; 4: PZE-101000083 anthesis   1 239225  0.503 0.298  -0.16     0.18    0.58</span></span>
<span><span class="co">#&gt; 5: PZE-101000108 anthesis   1 255850  0.654 0.185   0.17     0.34    0.90</span></span>
<span><span class="co">#&gt; 6: PZE-101000111 anthesis   1 263938  0.467 0.331  -0.41     0.24    0.83</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">GWAResult</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">[</span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">GWAResult</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">$</span><span class="va">trait</span> <span class="op">==</span> <span class="st">"grain.yield"</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;              snp       trait chr    pos pValue   LOD effect effectSe allFreq</span></span>
<span><span class="co">#&gt; 1:         SYN83 grain.yield   1   3498  0.928 0.033 -0.060    0.096    0.60</span></span>
<span><span class="co">#&gt; 2: PZE-101000060 grain.yield   1 157104  0.185 0.733  0.038    0.102    0.72</span></span>
<span><span class="co">#&gt; 3: PZE-101000088 grain.yield   1 238347  0.078 1.110  0.018    0.124    0.84</span></span>
<span><span class="co">#&gt; 4: PZE-101000083 grain.yield   1 239225  0.503 0.298 -0.110    0.091    0.58</span></span>
<span><span class="co">#&gt; 5: PZE-101000108 grain.yield   1 255850  0.654 0.185 -0.092    0.163    0.90</span></span>
<span><span class="co">#&gt; 6: PZE-101000111 grain.yield   1 263938  0.467 0.331 -0.031    0.119    0.83</span></span></code></pre></div>
<p>The <code>data.tables</code> above contain the results for traits
anthesis and grain.yield respectively. While the column effect is trait
specific, the p-Value is for the global null-hypothesis (<span class="math inline">\(\beta_1 = \dots = \beta_p = 0\)</span>) described
above; these P-values are repeated for each trait.</p>
<ol start="2" style="list-style-type: decimal">
<li>
<strong>signSnp</strong>: a list of <code>data.tables</code>, one
for each trial for which the analysis was run, containing the
significant SNPs. Optionally also the SNPs close to the significant SNPs
are included in the <code>data.table</code>. See <a href="#sign"><strong>Significance thresholds</strong></a> for more
information on how to do this. The data.tables in signSnp consist of the
same columns as those in GWAResult described above. Two extra columns
are added:</li>
</ol>
<table class="table">
<colgroup>
<col width="16%">
<col width="84%">
</colgroup>
<tbody>
<tr class="odd">
<td align="left">snpStatus</td>
<td align="left">either “significant SNP” or “within … of a significant
SNP”</td>
</tr>
<tr class="even">
<td align="left">propSnpVar</td>
<td align="left">proportion of the variance explained by the SNP,
computed as <span class="math inline">\(\beta_{\textrm{SNP}}^2 *
var(\textrm{SNP}) / var(\textrm{pheno})\)</span>
</td>
</tr>
</tbody>
</table>
<p>In this case there are no significant SNPs:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GWASDrops</span><span class="op">$</span><span class="va">signSnp</span><span class="op">$</span><span class="va">Mur13W</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li><p><strong>kinship</strong>: the kinship matrix (or matrices) used
in the GWAS analysis. This can either be the user provided kinship
matrix or the kinship matrix computed when running the
<code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function.</p></li>
<li><p><strong>thr</strong>: a list of thresholds, one for each trial
for which the analysis was run, used for determining significant
SNPs.</p></li>
<li><p><strong>GWASInfo</strong>: additional information on the
analysis, e.g. the call and the type of threshold used.</p></li>
</ol>
<div class="section level4">
<h4 id="gwas-summary">GWAS Summary<a class="anchor" aria-label="anchor" href="#gwas-summary"></a>
</h4>
<p>For a quick overview of the results, e.g. the number of significant
SNPs, use the summary function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create summary of GWASDrops for the trait grain number.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">GWASDrops</span>, traits <span class="op">=</span> <span class="st">"grain.number"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Mur13W:</span></span>
<span><span class="co">#&gt;  Traits analysed: anthesis, grain.number, grain.yield, plant.height, silking </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Data are available for 36624 SNPs.</span></span>
<span><span class="co">#&gt;   0 of them were not analyzed because their minor allele frequency is below 0.01 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  GLSMethod: single </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Trait: grain.number </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      LOD-threshold: 5.9 </span></span>
<span><span class="co">#&gt;      No significant SNPs found. </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      No genomic control correction was applied</span></span>
<span><span class="co">#&gt;      Genomic control inflation-factor: 1.7</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gwas-plots">GWAS Plots<a class="anchor" aria-label="anchor" href="#gwas-plots"></a>
</h4>
<p>The <code>plot</code> function can be used to visualize the results
in GWASDrops, with a QQ-plot, Manhattan plot or QTL-plot. More details
for each <code>plotType</code> are available in the
<code>statgenGWAS</code> vignette.</p>
<div class="section level5">
<h5 id="qq-plots">QQ plots<a class="anchor" aria-label="anchor" href="#qq-plots"></a>
</h5>
<p>A QQ-plot of the observed against the expected <span class="math inline">\(-\log_{10}(p)\)</span> values can be made by
setting <code>plotType = "qq"</code>. Most of the SNPs are expected to
have no effect, resulting in P-values uniformly distributed on <span class="math inline">\([0,1]\)</span>, and leading to the identity
function (<span class="math inline">\(y=x\)</span>) on the <span class="math inline">\(-\log_{10}(p)\)</span> scale. As in the plot
below, deviations from this line should only occur on the right side of
the plot, for a small number of SNPs with an effect on the phenotype
(and possibly SNPs in LD). There is <strong>inflation</strong> if the
observed <span class="math inline">\(-\log_{10}(p)\)</span> values are
always above the line <span class="math inline">\(y=x\)</span>, and
(less common) <strong>deflation</strong> if they are always below this
line. A QQ-plot therefore gives a first impression of the quality of the
GWAS model: if for example <span class="math inline">\(-\log_{10}(p)\)</span> values are consistently too
large (inflation), the correction for genetic relatedness may not be
adequate. In this case it may be of interest to correct the P-values for
inflation, using the <code>genomicControl</code> argument in the
<code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a qq plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qq"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/qqMtg-1.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="manhattan-plots">Manhattan plots<a class="anchor" aria-label="anchor" href="#manhattan-plots"></a>
</h5>
<p>A Manhattan plot is made by setting
<code>plotType = "manhattan"</code>. Significant SNPs are marked in
red.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a manhattan plot of GWAS Drops.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/manhattanMtg-1.png" width="672"></p>
<p>More options linked with <code>plotType = "manhattan"</code> are
described in the <code>statgenGWAS</code> vignette.</p>
</div>
<div class="section level5">
<h5 id="qtl-plots">QTL plots<a class="anchor" aria-label="anchor" href="#qtl-plots"></a>
</h5>
<p>A qtl plot can be made by setting <code>plotType = "qtl"</code>. In
this plot the significant SNPs are marked by circles at their genomic
positions, with diameter proportional to the estimated effect size; for
an example see <span class="citation">Millet et al. (<a href="#ref-Millet2016" role="doc-biblioref">2016</a>)</span>. Typically,
this is done for multiple traits or environments, with the genomic
position on the x-axis, which are displayed horizontally above each
other and can thus be compared.</p>
<p>Since the traits are measured on a different scale, the effect
estimates cannot be compared directly. For better comparison, one can
set <code>normalize = TRUE</code>, which divides the estimates by the
standard deviation of the phenotype.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot a qtl plot of GWAS Drops for Mur13W.</span></span>
<span><span class="co">## Set significance threshold to 5 and normalize effect estimates.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDrops</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span>, yThr <span class="op">=</span> <span class="fl">5</span>, normalize <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/qtlMtgNorm-1.png" width="672"></p>
<p>Other arguments can be used to plot a subset of the chromosomes
(<code>chr</code>) and directly export the plot to .pptx
(<code>exportPptx = TRUE</code> and specify <code>pptxName</code>). Note
that the <a href="https://ardata-fr.github.io/officeverse/" class="external-link">officer</a>
package is required for this. A full list of arguments can be found by
running <code>help(plot.GWAS)</code>.</p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kinship-matrices">Kinship matrices<a class="anchor" aria-label="anchor" href="#kinship-matrices"></a>
</h3>
<p>The <code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function has an argument
<code>kinshipMethod</code>, which defines the kinship matrix used for
association mapping. Kinship matrices can be computed directly using the
<code>kinship</code> function or within the
<code>runMultiTraitGwas function</code>. There are five options: (1)
using the covariance between the scaled SNP-scores
(<code>kinshipMethod = "astle"</code>, the default; see e.g. equation
(2.2) in <span class="citation">Astle and Balding (<a href="#ref-Astle2009" role="doc-biblioref">2009</a>)</span>) (2)
Identity by State (<code>kinshipMethod = "IBS"</code>; see e.g. equation
(2.3) in <span class="citation">Astle and Balding (<a href="#ref-Astle2009" role="doc-biblioref">2009</a>)</span>) (3) using
the formula by <span class="citation">VanRaden (<a href="#ref-VanRaden2008" role="doc-biblioref">2008</a>)</span>
(<code>kinshipMethod = "vanRaden"</code>) (4) using an identity matrix
(<code>kinshipMethod = "identity"</code>), corresponding to no kinship
correction (5) User-defined, in which case the argument <code>kin</code>
needs to be specified.</p>
<p>By default, the same kinship matrix is used for testing all SNPs
(<code>GLSMethod = "single"</code>). When
<code>GLSMethod = "multi"</code>, the kinship matrix is
chromosome-specific. In this case, the function fits variance components
and computes effect-estimates and P-values for each chromosome in turn,
using the kinship matrix for that chromosome (i.e. using all SNPs that
are <strong>not</strong> on this chromosome). Each chromosome-specific
kinship matrix is computed using the method specified by the argument
<code>kinshipMethod</code>. As shown by <span class="citation">Rincent
et al. (<a href="#ref-Rincent2014" role="doc-biblioref">2014</a>)</span>, this can give a considerable
improvement in power.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for trial 'Mur13W'.</span></span>
<span><span class="co">## Use chromosome specific kinship matrices computed using method of van Raden.</span></span>
<span><span class="va">GWASDropsChrSpec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>, </span>
<span>                                      trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                      GLSMethod <span class="op">=</span> <span class="st">"multi"</span>,</span>
<span>                                      kinshipMethod <span class="op">=</span> <span class="st">"vanRaden"</span>,</span>
<span>                                      covModel <span class="op">=</span> <span class="st">"fa"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="worked-example-2-one-trait-measured-in-multiple-trials">Worked example 2: one trait measured in multiple trials<a class="anchor" aria-label="anchor" href="#worked-example-2-one-trait-measured-in-multiple-trials"></a>
</h2>
<div class="section level3">
<h3 id="create-gdata-object-1">
<strong>Create gData object</strong><a class="anchor" aria-label="anchor" href="#create-gdata-object-1"></a>
</h3>
<p>In this example, we will focus on one trait, grain yield, in all
trials. <code>dropsPheno</code> contains genotypic means for 10 trials.
To be able to run a GWAS analysis with one trait measured in all trials,
the data has to be reshaped and added as a single data.frame with
“genotype” as first column and traits after that.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Reshape phenotypic data to data.frame in wide format containing only grain.yield.</span></span>
<span><span class="va">phenoDat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/reshape.html" class="external-link">reshape</a></span><span class="op">(</span><span class="va">dropsPheno</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Experiment"</span>,<span class="st">"genotype"</span>,<span class="st">"grain.yield"</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                    timevar <span class="op">=</span> <span class="st">"Experiment"</span>, </span>
<span>                    idvar <span class="op">=</span> <span class="st">"genotype"</span>, </span>
<span>                    direction <span class="op">=</span> <span class="st">"wide"</span>, </span>
<span>                    v.names <span class="op">=</span> <span class="st">"grain.yield"</span><span class="op">)</span></span>
<span><span class="co">## Rename columns to trial name only.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">phenoDat</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">phenoDat</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"grain.yield."</span>, replacement <span class="op">=</span> <span class="st">""</span>,</span>
<span>       x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">phenoDat</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">phenoDat</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Now we create a <code>gData</code> object containing map marker
information and phenotypes.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create a gData object containing map, marker and phenotypic information.</span></span>
<span><span class="va">gDataDropsxE</span> <span class="op">&lt;-</span> <span class="fu">statgenGWAS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/statgenGWAS/man/gData.html" class="external-link">createGData</a></span><span class="op">(</span>geno <span class="op">=</span> <span class="va">dropsMarkers</span>,</span>
<span>                                         map <span class="op">=</span> <span class="va">dropsMap</span>, </span>
<span>                                         pheno <span class="op">=</span> <span class="va">phenoDat</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gDataDropsxE</span><span class="op">)</span></span>
<span><span class="co">#&gt; map</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of chromosomes: 10 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; markers</span></span>
<span><span class="co">#&gt;  Number of markers: 41722 </span></span>
<span><span class="co">#&gt;  Number of genotypes: 246 </span></span>
<span><span class="co">#&gt;  Content:</span></span>
<span><span class="co">#&gt;   0 1 2 NA </span></span>
<span><span class="co">#&gt;   0.28 0.01 0.71 0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; pheno</span></span>
<span><span class="co">#&gt;  Number of trials: 1 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  phenoDat:</span></span>
<span><span class="co">#&gt;      Number of traits: 10 </span></span>
<span><span class="co">#&gt;      Number of genotypes: 246 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;         Kar12W Gai12W Kar13W Ner12R Mar13R Mur13R Cra12R Cam12R Kar13R Mur13W</span></span>
<span><span class="co">#&gt; Min.       5.4    7.5    4.1    2.1    3.8    2.1  0.088   0.38    5.6    3.3</span></span>
<span><span class="co">#&gt; 1st Qu.    8.8   10.4    7.0    4.0    7.0    5.8  0.891   1.33    8.7    6.3</span></span>
<span><span class="co">#&gt; Median     9.9   11.2    7.9    4.7    7.7    6.9  1.381   1.87    9.8    7.5</span></span>
<span><span class="co">#&gt; Mean       9.7   11.2    8.0    4.7    7.8    6.9  1.492   1.98    9.9    7.4</span></span>
<span><span class="co">#&gt; 3rd Qu.   10.7   12.0    9.1    5.3    8.6    7.8  1.971   2.59   11.0    8.4</span></span>
<span><span class="co">#&gt; Max.      13.1   14.3   12.7    7.1   11.5   10.6  4.979   4.90   13.8   11.4</span></span>
<span><span class="co">#&gt; NA's       0.0    0.0    0.0    0.0    0.0    0.0  0.000   0.00    0.0    0.0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="recoding-and-cleaning-of-markers-1">
<strong>Recoding and cleaning of markers</strong><a class="anchor" aria-label="anchor" href="#recoding-and-cleaning-of-markers-1"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Remove duplicate SNPs from gDataDrops.</span></span>
<span><span class="va">gDataDropsDedupxE</span> <span class="op">&lt;-</span> <span class="fu">statgenGWAS</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/statgenGWAS/man/codeMarkers.html" class="external-link">codeMarkers</a></span><span class="op">(</span><span class="va">gDataDropsxE</span>, impute <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                              verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">#&gt; Input contains 41722 SNPs for 246 genotypes.</span></span>
<span><span class="co">#&gt; 0 genotypes removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 0 SNPs removed because proportion of missing values larger than or equal to 1.</span></span>
<span><span class="co">#&gt; 5098 duplicate SNPs removed.</span></span>
<span><span class="co">#&gt; Output contains 36624 SNPs for 246 genotypes.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="multi-trial-gwas">
<strong>Multi-trial GWAS</strong><a class="anchor" aria-label="anchor" href="#multi-trial-gwas"></a>
</h3>
<p>Similar to the first example we run a multi-trial GWAs using a factor
analytic model.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trial GWAS for one trait in all trials.</span></span>
<span><span class="va">GWASDropsxE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedupxE</span>, </span>
<span>                                 covModel <span class="op">=</span> <span class="st">"fa"</span><span class="op">)</span></span></code></pre></div>
<p>Among the significant SNPs we find the large QTL on chromosome 6
reported in <span class="citation">Millet et al. (<a href="#ref-Millet2016" role="doc-biblioref">2016</a>)</span>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDropsxE</span><span class="op">$</span><span class="va">signSnp</span><span class="op">$</span><span class="va">pheno</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt;              snp  trait chr      pos  pValue  LOD effect effectSe allFreq       snpStatus propSnpVar</span></span>
<span><span class="co">#&gt; 1:      SYN25281 Kar12W   6 18646369 1.6e-07  6.8   0.23    0.100    0.79 significant SNP      0.017</span></span>
<span><span class="co">#&gt; 2: PZE-106021363 Kar12W   6 18846283 7.4e-11 10.1   0.28    0.091    0.70 significant SNP      0.034</span></span>
<span><span class="co">#&gt; 3: PZE-106021410 Kar12W   6 18990291 2.5e-11 10.6   0.30    0.091    0.70 significant SNP      0.037</span></span>
<span><span class="co">#&gt; 4: PZE-106021419 Kar12W   6 18991091 1.5e-12 11.8   0.26    0.092    0.74 significant SNP      0.026</span></span>
<span><span class="co">#&gt; 5: PZE-106021420 Kar12W   6 18991117 2.2e-10  9.6   0.27    0.091    0.70 significant SNP      0.032</span></span>
<span><span class="co">#&gt; 6: PZE-106021424 Kar12W   6 18991481 4.2e-12 11.4   0.24    0.092    0.74 significant SNP      0.022</span></span></code></pre></div>
<div class="section level4">
<h4 id="gwas-summary-1">GWAS Summary<a class="anchor" aria-label="anchor" href="#gwas-summary-1"></a>
</h4>
<p>For a quick overview of the results, e.g. the number of significant
SNPs, we again use the summary function. We restrict the output to two
trials using the <code>traits</code> argument.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">GWASDropsxE</span>, traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mur13W"</span>, <span class="st">"Kar12W"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; phenoDat:</span></span>
<span><span class="co">#&gt;  Traits analysed: Cam12R, Cra12R, Gai12W, Kar12W, Kar13R, Kar13W, Mar13R, Mur13R, Mur13W, Ner12R </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Data are available for 36624 SNPs.</span></span>
<span><span class="co">#&gt;   0 of them were not analyzed because their minor allele frequency is below 0.01 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  GLSMethod: single </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Trait: Mur13W </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      LOD-threshold: 5.9 </span></span>
<span><span class="co">#&gt;      Number of significant SNPs: 8 </span></span>
<span><span class="co">#&gt;      Smallest p-value among the significant SNPs: 1.5e-12 </span></span>
<span><span class="co">#&gt;      Largest p-value among the significant SNPs: 3.9e-07 (LOD-score: 6.4)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      No genomic control correction was applied</span></span>
<span><span class="co">#&gt;      Genomic control inflation-factor: 0.94 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Trait: Kar12W </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      LOD-threshold: 5.9 </span></span>
<span><span class="co">#&gt;      Number of significant SNPs: 8 </span></span>
<span><span class="co">#&gt;      Smallest p-value among the significant SNPs: 1.5e-12 </span></span>
<span><span class="co">#&gt;      Largest p-value among the significant SNPs: 3.9e-07 (LOD-score: 6.4)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      No genomic control correction was applied</span></span>
<span><span class="co">#&gt;      Genomic control inflation-factor: 0.94</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gwas-plots-1">GWAS Plots<a class="anchor" aria-label="anchor" href="#gwas-plots-1"></a>
</h4>
<p>As in the first example we use the <code>plot.GWAS()</code> function
to visualize the results in <code>GWASDropsxE</code>, with a QQ-plot,
Manhattan plot or QTL-plot.</p>
<div class="section level5">
<h5 id="qq-plots-1">QQ plots<a class="anchor" aria-label="anchor" href="#qq-plots-1"></a>
</h5>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDropsxE</span>, plotType <span class="op">=</span> <span class="st">"qq"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/qqMtgxE-1.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="manhattan-plots-1">Manhattan plots<a class="anchor" aria-label="anchor" href="#manhattan-plots-1"></a>
</h5>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDropsxE</span>, plotType <span class="op">=</span> <span class="st">"manhattan"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/manhattanMtgxE-1.png" width="672"></p>
</div>
<div class="section level5">
<h5 id="qtl-plots-1">QTL plots<a class="anchor" aria-label="anchor" href="#qtl-plots-1"></a>
</h5>
<p>The trait is measured with the same scale across trials so the effect
estimates can be compared directly (one can set
<code>normalize = FALSE</code>).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set significance threshold to 6 and do not normalize effect estimates.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">GWASDropsxE</span>, plotType <span class="op">=</span> <span class="st">"qtl"</span>, yThr <span class="op">=</span> <span class="fl">6</span>, normalize <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="statgenQTLxT_files/figure-html/qtlMtgNormxE-1.png" width="672"></p>
<hr>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="opts">Further options<a class="anchor" aria-label="anchor" href="#opts"></a>
</h2>
<p>The <code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function has many more arguments
that can be specified. In this section similar arguments are grouped and
explained with examples on how to use them.</p>
<div class="section level3">
<h3 id="sign">Significance thresholds<a class="anchor" aria-label="anchor" href="#sign"></a>
</h3>
<p>The threshold for selecting significant SNPs in a GWAS analysis is
computed by default using Bonferroni correction, with an alpha of 0.05.
The alpha can be modified setting the option <em>alpha</em> when calling
<code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code>. Two other threshold types can be used:
a fixed threshold (<code>thrType = "fixed"</code>) specifying the <span class="math inline">\(-\log_{10}(p)\)</span> (<code>LODThr</code>) value
of the threshold, or a threshold that defines the n SNPs with the
highest <span class="math inline">\(-\log_{10}(p)\)</span> scores as
significant SNPs. Set <code>thrType = "small"</code> together with
<code>nSnpLOD = n</code> to do this. In the following example, we define
all SNPs with <span class="math inline">\(p &lt; 10^{-4}\)</span> as
significant SNPs.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for Mur13W.</span></span>
<span><span class="co">## Use a fixed significance threshold of 4.</span></span>
<span><span class="va">GWASDropsFixThr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                     trials <span class="op">=</span> <span class="st">"Mur13W"</span>, </span>
<span>                                     covModel <span class="op">=</span> <span class="st">"fa"</span>,</span>
<span>                                     thrType <span class="op">=</span> <span class="st">"fixed"</span>,</span>
<span>                                     LODThr <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<p><strong>Controlling false discovery rate</strong></p>
<p>A final option for selecting significant SNPs is by setting
<code>thrType = "fdr"</code>. When doing so the significant SNPs won’t
be selected by computing a genome wide threshold, but by trying to
control the rate of false discoveries as in <span class="citation">Brzyski et al. (<a href="#ref-Brzyski2016" role="doc-biblioref">2016</a>)</span>.</p>
<p>First, a list is defined containing all SNPs with a P-value below
<code>pThr</code>, default 0.05. Then clusters of SNPs are created using
a two step iterative process in which SNPs with the lowest P-values are
selected as cluster representatives. This SNP and all SNPs that have a
correlation with this SNP of <span class="math inline">\(\rho\)</span>
or higher (specified by the function argument <code>rho</code>, default
0.4) will form a cluster. The selected SNPs are removed from the list
and the procedure is repeated until no SNPs are left. At the end of this
step, one has a list of clusters, with corresponding vector of P-values
of the cluster representatives. Finally, to determine the number of
significant clusters, the first cluster is determined for which the
P-value of the cluster representative is larger than <span class="math inline">\(cluster_{number} * \alpha / m\)</span>, where
<span class="math inline">\(m\)</span> is the number of SNPs and <span class="math inline">\(\alpha\)</span> can be specified by the
corresponding function argument. All previous clusters are selected as
significant.</p>
</div>
<div class="section level3">
<h3 id="variance-components">Variance components<a class="anchor" aria-label="anchor" href="#variance-components"></a>
</h3>
<p>There are three ways to compute the variance components used in the
GWAS analysis. These can be specified by setting the argument
<code>covModel</code>. See <a href="#vc"><strong>Models for the genetic
and residual covariance</strong></a> for a description of the
options.</p>
<p><strong>Note that <code>covModel = unst</code> can only be used for
less than 10 traits or trials. It is not recommended to use it for six
to nine trials for computational reasons.</strong></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for for Mur13W.</span></span>
<span><span class="co">## Use a factor analytic model for computing the variance components.</span></span>
<span><span class="va">GWASDropsFA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                 trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                 covModel <span class="op">=</span> <span class="st">"fa"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Rerun the analysis, using the variance components computed in the </span></span>
<span><span class="co">## previous model as inputs.</span></span>
<span><span class="va">GWASDropsFA2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                  trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                  fitVarComp  <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                  Vg <span class="op">=</span> <span class="va">GWASDropsFA</span><span class="op">$</span><span class="va">GWASInfo</span><span class="op">$</span><span class="va">varComp</span><span class="op">$</span><span class="va">Vg</span>,</span>
<span>                                  Ve <span class="op">=</span> <span class="va">GWASDropsFA</span><span class="op">$</span><span class="va">GWASInfo</span><span class="op">$</span><span class="va">varComp</span><span class="op">$</span><span class="va">Ve</span><span class="op">)</span></span></code></pre></div>
<p><strong>Parallel computing</strong></p>
<p>To improve performance when using a pairwise variance component
model, it is possible to use parallel computing. To do this, a parallel
back-end has to be specified by the user, e.g. by using
<code>registerDoParallel</code> from the <code>doParallel</code> package
(see the example below). In addition, in the
<code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function the argument
<code>parallel</code> has to be set to <code>TRUE</code>. With these
settings the computations are done in parallel per pair of traits.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Register parallel back-end with 2 cores.</span></span>
<span><span class="fu">doParallel</span><span class="fu">::</span><span class="fu">registerDoParallel</span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run multi-trait GWAS for one trait in all trials.</span></span>
<span><span class="va">GWASDropsxEPar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedupxE</span>,</span>
<span>                                    covModel <span class="op">=</span> <span class="st">"pw"</span>,</span>
<span>                                    parallel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="covariates">Covariates<a class="anchor" aria-label="anchor" href="#covariates"></a>
</h3>
<p>Covariates can be included as extra fixed effects in the GWAS model.
The <code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> function distinguishes between
‘usual’ covariates and SNP-covariates. The former could be design
factors such as block, or other traits one wants to condition on. In the
latter case, the covariate(s) are one or more of the markers contained
in the genotypic data. SNP-covariates can be set with the argument
<code>snpCov</code>, which should be a vector of marker names.
Similarly, other covariates should be specified using the argument
<code>covar</code>, containing a vector of covariate names. The
<code>gData</code> object should contain these covariates in
<code>gData$covar</code>.</p>
<p>In case SNP-covariates are used, GWAS for all the other SNPs is
performed with the the SNP-covariates as extra fixed effect; also the
null model used to estimate the variance components includes these
effects. For each SNP in SNP-covariates, a P-value is obtained using the
same F-test and null model to estimate the variance components, but with
only all other SNPs (if any) in SNP-covariates as fixed effects.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for Mur13W.</span></span>
<span><span class="co">## Use PZE-106021410, the most significant SNP, as SNP covariate.</span></span>
<span><span class="va">GWASDropsSnpCov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                     trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                     snpCov <span class="op">=</span> <span class="st">"PZE-106021410"</span>,</span>
<span>                                     covModel <span class="op">=</span> <span class="st">"fa"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="minor-allele-frequency">Minor Allele Frequency<a class="anchor" aria-label="anchor" href="#minor-allele-frequency"></a>
</h3>
<p>It is recommended to remove SNPs with a low minor allele frequency
(MAF) from the data before starting a GWAS analysis. However it is also
possible to do so in the analysis itself. The difference between these
approaches is that <code>codeMarkers()</code> removes the SNPs, whereas
<code><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas()</a></code> excludes them from the analysis but
leaves them in the output (with results set to <code>NA</code>). In the
latter case it will still be possible to see the allele frequency of the
SNP.<br>
By default all SNPs with a MAF lower than 0.01 are excluded from the
analysis. This can be controlled by the argument <code>MAF</code>.
Setting MAF to 0 will still exclude duplicate SNPs since duplicates
cause problems when fitting the underlying models.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for Mur13W.</span></span>
<span><span class="co">## Only include SNPs that have a MAF of 0.05 or higher.</span></span>
<span><span class="va">GWASDropsMAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                  trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                  covModel <span class="op">=</span> <span class="st">"fa"</span>,</span>
<span>                                  MAF <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimation-of-common-snp-effects-and-qtle-effects-">Estimation of common SNP effects and QTL×E effects.<a class="anchor" aria-label="anchor" href="#estimation-of-common-snp-effects-and-qtle-effects-"></a>
</h3>
<p>Besides a normal SNP-effect model, it is possible to fit a common
SNP-effect model as well (see <a href="#common"><strong>Hypotheses for
the SNP-effects</strong></a>). When doing so, in addition to the
SNP-effect, also the common SNP-effect and the QTL×E effect and
corresponding standard errors and P-values are returned. These are
included as extra columns in the GWAResult <code>data.table</code> in
the output of the function.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run multi-trait GWAS for Mur13W.</span></span>
<span><span class="co">## Fit an additional common sNP effect model.</span></span>
<span><span class="va">GWASDropsCommon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runMultiTraitGwas.html">runMultiTraitGwas</a></span><span class="op">(</span>gData <span class="op">=</span> <span class="va">gDataDropsDedup</span>,</span>
<span>                                     trials <span class="op">=</span> <span class="st">"Mur13W"</span>,</span>
<span>                                     covModel <span class="op">=</span> <span class="st">"fa"</span>,</span>
<span>                                     estCom <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GWASDropsCommon</span><span class="op">$</span><span class="va">GWAResult</span><span class="op">$</span><span class="va">Mur13W</span><span class="op">)</span></span>
<span><span class="co">#&gt;              snp    trait chr    pos pValue    LOD effect effectSe allFreq pValCom effsCom effsComSe pValQtlE</span></span>
<span><span class="co">#&gt; 1:         SYN83 anthesis   1   3498   0.98 0.0087  -0.12     0.19    0.60   0.536  -0.028     0.045     0.98</span></span>
<span><span class="co">#&gt; 2: PZE-101000060 anthesis   1 157104   0.20 0.6991   0.47     0.20    0.72   0.029   0.103     0.048     0.58</span></span>
<span><span class="co">#&gt; 3: PZE-101000088 anthesis   1 238347   0.19 0.7142  -0.61     0.25    0.84   0.049  -0.115     0.059     0.44</span></span>
<span><span class="co">#&gt; 4: PZE-101000083 anthesis   1 239225   0.76 0.1207  -0.16     0.18    0.58   0.203  -0.054     0.043     0.88</span></span>
<span><span class="co">#&gt; 5: PZE-101000108 anthesis   1 255850   0.60 0.2253   0.15     0.34    0.90   0.774  -0.023     0.079     0.48</span></span>
<span><span class="co">#&gt; 6: PZE-101000111 anthesis   1 263938   0.60 0.2217  -0.37     0.24    0.83   0.248  -0.064     0.055     0.66</span></span></code></pre></div>
<p>The SNPs with a significant QTLxE effect come from the same regions
as the SNPs in supplementary table 6 in <span class="citation">Millet et
al. (<a href="#ref-Millet2016" role="doc-biblioref">2016</a>)</span>,
who however used a different SNP set. Another difference is that here we
explicitly test for QTLxE within the GWAS, whereas <span class="citation">Millet et al. (<a href="#ref-Millet2016" role="doc-biblioref">2016</a>)</span> regressed SNP effects on
environmental covariates.</p>
<hr>
</div>
<div class="section level3">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Astle2009" class="csl-entry">
Astle, William, and David J. Balding. 2009. <span>“Population Structure
and Cryptic Relatedness in Genetic Association Studies.”</span>
<em>Statistical Science</em> 24 (4): 451–71. <a href="https://doi.org/10.1214/09-sts307" class="external-link">https://doi.org/10.1214/09-sts307</a>.
</div>
<div id="ref-Brzyski2016" class="csl-entry">
Brzyski, Damian, Christine B. Peterson, Piotr Sobczyk, Emmanuel J.
Candès, Malgorzata Bogdan, and Chiara Sabatti. 2016. <span>“Controlling
the Rate of <span>GWAS</span> False Discoveries.”</span>
<em>Genetics</em> 205 (1): 61–75. <a href="https://doi.org/10.1534/genetics.116.193987" class="external-link">https://doi.org/10.1534/genetics.116.193987</a>.
</div>
<div id="ref-Furlotte2015" class="csl-entry">
Furlotte, N. A., and E. Eskin. 2015. <span>“Efficient Multiple-Trait
Association and Estimation of Genetic Correlation Using the
Matrix-Variate Linear Mixed Model.”</span> <em>Genetics</em> 200 (1):
59–68. <a href="https://doi.org/10.1534/genetics.114.171447" class="external-link">https://doi.org/10.1534/genetics.114.171447</a>.
</div>
<div id="ref-Korte2012" class="csl-entry">
Korte, Arthur, Bjarni J Vilhjálmsson, Vincent Segura, Alexander Platt,
Quan Long, and Magnus Nordborg. 2012. <span>“A Mixed-Model Approach for
Genome-Wide Association Studies of Correlated Traits in Structured
Populations.”</span> <em>Nature Genetics</em> 44 (9): 1066–71. <a href="https://doi.org/10.1038/ng.2376" class="external-link">https://doi.org/10.1038/ng.2376</a>.
</div>
<div id="ref-Millet2019" class="csl-entry">
Millet, Emilie J., Cyril Pommier, Mélanie Buy, Axel Nagel, Willem
Kruijer, Therese Welz-Bolduan, Jeremy Lopez, et al. 2019. <span>“A
Multi-Site Experiment in a Network of European Fields for Assessing the
Maize Yield Response to Environmental Scenarios.”</span> Portail Data
Inra. <a href="https://doi.org/10.15454/IASSTN" class="external-link">https://doi.org/10.15454/IASSTN</a>.
</div>
<div id="ref-Millet2016" class="csl-entry">
Millet, Emilie J., Claude Welcker, Willem Kruijer, Sandra Negro,
Stephane Nicolas, Sebastien Praud, Nicolas Ranc, et al. 2016.
<span>“Genome-Wide Analysis of Yield in Europe: Allelic Effects as
Functions of Drought and Heat Scenarios.”</span> <em>Plant
Physiology</em>, July, pp.00621.2016. <a href="https://doi.org/10.1104/pp.16.00621" class="external-link">https://doi.org/10.1104/pp.16.00621</a>.
</div>
<div id="ref-Rincent2014" class="csl-entry">
Rincent, Renaud, Laurence Moreau, Hervé Monod, Estelle Kuhn, Albrecht E.
Melchinger, Rosa A. Malvar, Jesus Moreno-Gonzalez, et al. 2014.
<span>“Recovering Power in Association Mapping Panels with Variable
Levels of Linkage Disequilibrium.”</span> <em>Genetics</em> 197 (1):
375–87. <a href="https://doi.org/10.1534/genetics.113.159731" class="external-link">https://doi.org/10.1534/genetics.113.159731</a>.
</div>
<div id="ref-VanRaden2008" class="csl-entry">
VanRaden, P. M. 2008. <span>“Efficient Methods to Compute Genomic
Predictions.”</span> <em>Journal of Dairy Science</em> 91 (11): 4414–23.
<a href="https://doi.org/10.3168/jds.2007-0980" class="external-link">https://doi.org/10.3168/jds.2007-0980</a>.
</div>
<div id="ref-Zhou2014" class="csl-entry">
Zhou, Xiang, and Matthew Stephens. 2014. <span>“Efficient Multivariate
Linear Mixed Model Algorithms for Genome-Wide Association
Studies.”</span> <em>Nature Methods</em> 11 (4): 407–9. <a href="https://doi.org/10.1038/nmeth.2848" class="external-link">https://doi.org/10.1038/nmeth.2848</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bart-Jan van Rossum, Willem Kruijer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
