# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

updateFAHomVarCPP <- function(s, wNew, pNew, m, maxDiag = 1e4) {
    invisible(.Call(`_genStatPipeline_updateFAHomVarCPP`, s, wNew, pNew, m, maxDiag))
}

updateFACPP <- function(y, wStart, pStart, wNew, pNew, m0, hetVar = FALSE, maxDiag = 1e4, tolerance = 1e-4, maxIter = 100L, printProgress = FALSE) {
    invisible(.Call(`_genStatPipeline_updateFACPP`, y, wStart, pStart, wNew, pNew, m0, hetVar, maxDiag, tolerance, maxIter, printProgress))
}

updatePrecCPP <- function(m, nc, omega, w, p, het, maxDiag) {
    .Call(`_genStatPipeline_updatePrecCPP`, m, nc, omega, w, p, het, maxDiag)
}

#' Compute tYPY as in Zhou and Stephens eqn. 50.
#'
#' Compute \eqn{t(y) * P * y}, part of the log-likelihood functions from
#' equation 26 and 27 in Zhou and Stephens using equation 50. Equation 56, 57
#' and 58 are used to do the actual computations.
#'
#' It is assumed that X and Y have already been rotated by Uk, where Uk is such
#' that the kinship matrix K equals \eqn{K = Uk * Dk * t(Uk)}.\cr
#' The original X and Y are right multiplied by Uk, e.g. \code{Y <- Y * Uk}.
#' See Zhou and Stephens (2014), supplement.\cr
#' It is these rotated versions that are the input of this function.
#'
#' @inheritParams LLDiag
#'
#' @return A numerical value for the \eqn{t(y) * P * y} part of the
#' log-likelihood function.
#'
#' @references Zhou, X. and Stephens, M. (2014). Efficient multivariate linear
#' mixed model algorithms for genome-wide association studies. Nature Methods,
#' February 2014, Vol. 11, p. 407–409
#'
#' @keywords internal
LLQuadFormDiagCPP <- function(y, vInv, size_param = NULL) {
    .Call(`_genStatPipeline_LLQuadFormDiagCPP`, y, vInv, size_param)
}

#' Estimates for covariates
#'
#' Compute the estimates and standard errors for the covariates in the input
#' matrix W.
#'
#' @param y An n x p matrix of observed phenotypes, on p traits or environments
#' for n genotypes. No missing values are allowed.
#' @param w An n x c covariate matrix, c being the number of covariates and n
#' being the number of genotypes. c has to be at least one (typically an
#' intercept). No missing values are allowed.
#' @param x An n x ns matrix of marker scores. Neither missing values nor
#' non-segregating markers are allowed.
#' @param vg A p x p matrix of genetic covariances.
#' @param ve A p x p matrix of environmental covariances.
#' @param k An n x n genetic relatedness matrix.
#' @param returnSe Should standard errors and p-values be returned?
#' @param estCom Should the common SNP-effect model be fitted?
#' @param ncores An integer indicating the number of cores used for parallel
#' computation.
#'
#' @return A list containing the estimates, optionally the standard errors of
#' the estimates and corresponding p-values. If \code{estCom = TRUE} also
#' common SNP-effects, their standard errors and corresponding p-values and
#' the p-values for QtlxE are output.
#'
#' @references Zhou, X. and Stephens, M. (2014). Efficient multivariate linear
#' mixed model algorithms for genome-wide association studies. Nature Methods,
#' February 2014, Vol. 11, p. 407–409
estEffsCPP <- function(y, w, x, vg, ve, k, returnSe = TRUE, estCom = FALSE, ncores = 4L) {
    .Call(`_genStatPipeline_estEffsCPP`, y, w, x, vg, ve, k, returnSe, estCom, ncores)
}

fastGLSCPP <- function(X, y, sigma, size_param = NULL, ncores = 1L) {
    .Call(`_genStatPipeline_fastGLSCPP`, X, y, sigma, size_param, ncores)
}

fastGLSIBDCPP <- function(mp, y, sigma, ref, size_param = NULL, ncores = 1L) {
    .Call(`_genStatPipeline_fastGLSIBDCPP`, mp, y, sigma, ref, size_param, ncores)
}

astleCPP <- function(x, denom = NULL) {
    .Call(`_genStatPipeline_astleCPP`, x, denom)
}

GRMCPP <- function(x, denom = NULL) {
    .Call(`_genStatPipeline_GRMCPP`, x, denom)
}

IBSCPP <- function(x, denom = NULL) {
    .Call(`_genStatPipeline_IBSCPP`, x, denom)
}

vanRadenCPP <- function(x, denom = NULL) {
    .Call(`_genStatPipeline_vanRadenCPP`, x, denom)
}

multiAllKinCPP <- function(x, posCor, denom = NULL) {
    .Call(`_genStatPipeline_multiAllKinCPP`, x, posCor, denom)
}

