<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Perform multi-trait GWAS</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for runMultiTraitGwas {genStatPipeline}"><tr><td>runMultiTraitGwas {genStatPipeline}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Perform multi-trait GWAS</h2>

<h3>Description</h3>

<p><code>runMultiTraitGwas</code> performs a multi-trait Genome Wide Association
Study (GWAS) on phenotypic and genotypic data contained in a <code>gData</code>
object.
</p>


<h3>Usage</h3>

<pre>
runMultiTraitGwas(gData, environments = NULL, covar = NULL,
  snpCov = NULL, kin = NULL, kinshipMethod = c("astle", "GRM", "IBS",
  "vanRaden"), GLSMethod = c("single", "multi"), subsetMarkers = FALSE,
  markerSubset = "", MAF = 0.01, fitVarComp = TRUE,
  covModel = c("unst", "pw", "fa"), VeDiag = TRUE, tolerance = 1e-06,
  maxIter = 2e+05, maxDiag = 10000, mG = 1, mE = 1, CmHet = TRUE,
  DmHet = TRUE, stopIfDecreasing = TRUE, computeLogLik = TRUE,
  Vg = NULL, Ve = NULL, reduceK = FALSE, nPca = NULL,
  estCom = FALSE, parallel = FALSE, nCores = NULL)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>gData</code></td>
<td>
<p>An object of class <code>gData</code> containing at least <code>map</code>,
<code>markers</code> and <code>pheno</code>.</p>
</td></tr>
<tr valign="top"><td><code>environments</code></td>
<td>
<p>A vector of environments on which to run GWAS. These can
be either numeric indices or character names of list items in <code>pheno</code>.
If <code>NULL</code> GWAS is run for all environments. GWAS is run for the
selected environments in sequential order.</p>
</td></tr>
<tr valign="top"><td><code>covar</code></td>
<td>
<p>An optional vector of covariates taken into account when
running GWAS. These can be either numeric indices or character names of
columns in <code>covar</code> in <code>gData</code>. If <code>NULL</code> no covariates are
used.</p>
</td></tr>
<tr valign="top"><td><code>snpCov</code></td>
<td>
<p>An optional character vector of snps to be included as
covariates.</p>
</td></tr>
<tr valign="top"><td><code>kin</code></td>
<td>
<p>An optional kinship matrix or list of kinship matrices. These
matrices can be from the <code>matrix</code> class as defined in the base package
or from the <code>dsyMatrix</code> class, the class of symmetric matrices in the
Matrix package.<br />
If <code>GLSMethod</code> = &quot;single&quot; then one matrix should be provided, if
<code>GLSMethod</code> = &quot;multi&quot; a list of chromosome specific matrices of lenght
equal to the number of chromosomes in <code>map</code> in <code>gData</code>.<br />
If <code>NULL</code> then matrix <code>kinship</code> in <code>gData</code> is used. <br />
If both <code>kin</code> is provided and <code>gData</code> contains a matrix
<code>kinship</code> then <code>kin</code> is used.</p>
</td></tr>
<tr valign="top"><td><code>kinshipMethod</code></td>
<td>
<p>An optional character indicating the method used for
calculating the kinship matrix(ces). Currently &quot;astle&quot; (Astle and Balding,
2009), &quot;GRM&quot;, &quot;IBS&quot; and &quot;vanRaden&quot; (VanRaden, 2008) are supported. If a
kinship matrix is supplied either in <code>gData</code> or in parameter <code>kin</code>
<code>kinshipMethod</code> is ignored.</p>
</td></tr>
<tr valign="top"><td><code>GLSMethod</code></td>
<td>
<p>A character string indicating the method used to estimate
the marker effects. Either <code>single</code> for using a single kinship matrix.
or <code>multi</code> for using chromosome specific kinship matrices.</p>
</td></tr>
<tr valign="top"><td><code>subsetMarkers</code></td>
<td>
<p>Should the marker data be subsetted?</p>
</td></tr>
<tr valign="top"><td><code>markerSubset</code></td>
<td>
<p>A numeric or character vector used for subsetting the
markers. Ignored if subsetMarkers = <code>FALSE</code>.</p>
</td></tr>
<tr valign="top"><td><code>MAF</code></td>
<td>
<p>A numerical value between 0 and 1. SNPs with minor allele
frequency below this value are not taken into account for the analysis,
i.e. p-values and effect sizes are put to missing (NA). Ignored if
<code>useMAF</code> is <code>FALSE</code>.</p>
</td></tr>
<tr valign="top"><td><code>fitVarComp</code></td>
<td>
<p>Should the variance components be fitted? If <code>FALSE</code>
they should be supplied in Vg and Ve</p>
</td></tr>
<tr valign="top"><td><code>covModel</code></td>
<td>
<p>A character string indictating the model used when fitting
the variance components. Either <code>unst</code> for unstructured for both Vg and
Ve (as in Zhou and Stephens (2014)), <code>pw</code> for unstructered for both Vg
and Ve (pairwise, as in Furlotte and Eskin (2013)) or <code>fa</code> for
factor-analytic for both Vg and Ve.<br />
Ignored if <code>fitVarComp</code> = <code>FALSE</code></p>
</td></tr>
<tr valign="top"><td><code>VeDiag</code></td>
<td>
<p>Should there be environmental correlations if covModel = &quot;unst&quot;
or &quot;pw&quot;? If traits are measured on the same individuals put <code>FALSE</code>.</p>
</td></tr>
<tr valign="top"><td><code>tolerance</code></td>
<td>
<p>A numerical value. The iterating process stops if the
difference in conditional log-likelihood between two consecutive iterations
drops below <code>tolerance</code>. Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>maxIter</code></td>
<td>
<p>An integer for the maximum number of iterations. Only used
when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>maxDiag</code></td>
<td>
<p>A numical value. The maximal value of the diagonal elements
in the precision matrices Cm and Dm (ignoring the low-rank part W W^t).
Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>mG</code></td>
<td>
<p>An integer. The order of the genetic part of the factor analytic
model. Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>mE</code></td>
<td>
<p>An integer. The order of the environmental part of the factor
analytic model. Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>CmHet</code></td>
<td>
<p>Should an extra diagonal part be added in the model for the
precision matrix Cm? Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>DmHet</code></td>
<td>
<p>Should an extra diagonal part be added in the model for the
precision matrix Dm? Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>stopIfDecreasing</code></td>
<td>
<p>Should the iterating process in the factor analytic
model stop if after 50 iterations the log-likelihood decreases between two
consecutive iterations? Only used when <code>covModel = "fa"</code>.</p>
</td></tr>
<tr valign="top"><td><code>Vg</code></td>
<td>
<p>An optional matrix with genotypic variance components. Vg should
have row names column names corresponding to the column names of Y. It may
contain additional rows and colums which will be ignored. Ignored if
fitVarComp = <code>TRUE</code>.</p>
</td></tr>
<tr valign="top"><td><code>Ve</code></td>
<td>
<p>An optional matrix with environmental variance components. Ve
should have row names column names corresponding to the column names of Y.
It may contain additional rows and colums which will be ignored. Ignored if
fitVarComp = <code>TRUE</code>.</p>
</td></tr>
<tr valign="top"><td><code>reduceK</code></td>
<td>
<p>Should the kinship matrix be reduced? See
<code><a href="reduceKinship.html">reduceKinship</a></code></p>
</td></tr>
<tr valign="top"><td><code>nPca</code></td>
<td>
<p>An integer giving the number of Pcas used whe reducing the
kinship matrix. Ignored if reduceK = <code>FALSE</code>.</p>
</td></tr>
<tr valign="top"><td><code>estCom</code></td>
<td>
<p>Should the common SNP-effect model be fitted? If <code>TRUE</code>
not only the SNP-effects but also the common SNP-effect and QTL x E effect
are estimated.</p>
</td></tr>
<tr valign="top"><td><code>parallel</code></td>
<td>
<p>Should the computation of variance components be done in
parallel? Only used if <code>covModel = "pw"</code>. A parallel computing
environment has to be setup by the user.</p>
</td></tr>
<tr valign="top"><td><code>nCores</code></td>
<td>
<p>A numerical value indicating the number of cores to be used by
the parallel part of the algorithm. If <code>NULL</code> the number of cores used
will be equal to the number of cores available on the machine - 1.</p>
</td></tr>
</table>


<h3>Value</h3>

<p>An object of class <code><a href="GWAS.html">GWAS</a></code>.
</p>


<h3>References</h3>

<p>Dahl et al. (2013). Network inference in matrix-variate Gaussian
models with non-independent noise. arXiv preprint arXiv:1312.1622.
</p>
<p>Furlotte, N.A. and Eskin, E. (2015). Efficient multiple-trait
association and estimation of genetic correlation using the matrix-variate
linear mixed model. Genetics, May 2015, Vol.200-1, p. 59-68.
</p>
<p>Kruijer et al. (2015) Marker-based estimation of heritability in
immortal populations. Genetics. February 2015, Vol. 199-2, p. 379-398.
</p>
<p>Zhou, X. and Stephens, M. (2014). Efficient multivariate linear
mixed model algorithms for genome-wide association studies. Nature Methods,
February 2014, Vol. 11, p. 407â€“409.
</p>

<hr /><div style="text-align: center;">[Package <em>genStatPipeline</em> version 0.0.0.9000 <a href="00Index.html">Index</a>]</div>
</body></html>
